src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -0,0 +1,333 @@
+
+/**
+ * Represents a persistent allocation made in native memory. In this case, persistent means that the allocation is made
+ * and not freed in the same call to the JNI. Therefore, in order to prevent memory leaks, we need to ensure that each
+ * allocation is properly freed
+ */
+public interface NativeMemoryAllocation {
+
+    /**
+     * Closes the native memory allocation. It should deallocate all native memory associated with this allocation.
+     */
+    void close();
+
+    /**
+     * Check if the allocation has been closed.
+     *
+     * @return true if allocation has been closed; false otherwise
+     */
+    boolean isClosed();
+
+    /**
+     * Get the native memory pointer associated with the native memory allocation.
+     *
+     * @return pointer to native memory allocation
+     */
+    long getPointer();
+
+    /**
+     * Locks allocation for read. Multiple threads can obtain this lock assuming that no threads have the write lock.
+     */
+    void readLock();
+
+    /**
+     * Locks allocation for write. Only one thread can obtain this lock and no threads can have a read lock.
+     */
+    void writeLock();
+
+    /**
+     * Unlocks allocation for read.
+     */
+    void readUnlock();
+
+    /**
+     * Unlocks allocation for write.
+     */
+    void writeUnlock();
+
+    /**
+     * Get the size of the native memory allocation in kilobytes.
+     *
+     * @return size of native memory allocation
+     */
+    long getSizeInKb();
+
+    /**
+     * Represents native indices loaded into memory. Because these indices are backed by files, they should be
+     * freed when file is deleted.
+     */
+    class IndexAllocation implements NativeMemoryAllocation {
+
+        private final long pointer;
+        private final long size;
+        private volatile boolean closed;
+        private final KNNEngine knnEngine;
+        private final String indexPath;
+        private final String openSearchIndexName;
+        private final ReadWriteLock readWriteLock;
+        private final WatcherHandle<FileWatcher> watcherHandle;
+
+        /**
+         * Constructor
+         *
+         * @param pointer Pointer in memory to the index
+         * @param size Size this index consumes in kilobytes
+         * @param knnEngine KNNEngine associated with the index allocation
+         * @param indexPath File path to index
+         * @param openSearchIndexName Name of OpenSearch index this index is associated with
+         * @param watcherHandle Handle for watching index file
+         */
+        IndexAllocation(long pointer, long size, KNNEngine knnEngine, String indexPath,
+                               String openSearchIndexName, WatcherHandle<FileWatcher> watcherHandle) {
+            this.closed = false;
+            this.knnEngine = knnEngine;
+            this.indexPath = indexPath;
+            this.openSearchIndexName = openSearchIndexName;
+            this.pointer = pointer;
```

Comment:
Can this be made Long? I see that JNIService.load... returns long, having this object avoids != 0 check instead of != null. What do you think?